### ISSUE TITLE
HSTUModel does not learn (train_loss stays flat) while SASRecModel with identical hyperparameters trains normally

### ISSUE THREAD

**Issue by Sleptsov99** [MobileTeleSystems/RecTools#324]
### What happened?

HSTUModel does not appear to learn on our data — train_loss/val_loss stay
essentially flat over many epochs, while SASRecModel with identical
hyperparameters on the same tiny dataset trains normally (loss drops by >15x).

Minimal reproduction (100 users, sampled from a larger interactions log, to
rule out scale/hyperparameter issues):

from rectools.dataset import Dataset
from rectools.models import HSTUModel, SASRecModel
from rectools.models.nn.transformers.utils import leave_one_out_mask
from pytorch_lightning import Trainer

dataset = Dataset.construct(tiny_train_df)  # 100 users, standard user_id/item_id/datetime/weight columns

common_kwargs = dict(
    n_blocks=1, n_heads=1, n_factors=32, session_max_len=150,
    lr=0.01, batch_size=16, n_negatives=10, epochs=50, verbose=1,
)

# SASRec — trains fine
sasrec = SASRecModel(**common_kwargs,
    get_trainer_func=lambda: Trainer(max_epochs=50, accelerator="gpu", devices=1, logger=True))
sasrec.fit(dataset)
# train_loss: 7.53 -> 0.48 over 50 epochs

# HSTU — does not train
hstu = HSTUModel(**common_kwargs,
    get_val_mask_func=leave_one_out_mask,
    get_trainer_func=lambda: Trainer(max_epochs=50, accelerator="gpu", devices=1, logger=True))
hstu.fit(dataset)
# train_loss: 7.50 -> 6.70 over 50 epochs (essentially flat)

Isolation steps taken:

1. Ruled out data/scale issues: same behavior reproduced from 2% up to 50% of
   our full dataset (9.3M+ users, ~460M interactions over a 120-day window),
   and on this 100-user tiny sample where SASRec clearly overfits as expected.

2. Ruled out learning rate / batch size / n_negatives: tried lr up to 0.05,
   batch_size up to 1024, n_negatives up to 100 — no meaningful change in the
   loss curve (train_loss stays in the 6.6–7.0 range in all configurations).

3. Ruled out relative_time_attention / relative_pos_attention: set both to
   False (the only two HSTU-specific params not shared with SASRecModel,
   confirmed via diffing get_config(simple_types=True) between both model
   classes with identical explicit kwargs) — no change (loss still ~6.68–6.70).

4. Swapped transformer_layers_type=SASRecTransformerLayers while keeping the
   rest of the HSTU pipeline (data_preparator, lightning_module, backbone) —
   loss still ~6.68. Note: SASRecTransformerLayers.__init__ accepts **kwargs,
   so relative_*_attention are silently ignored here, but the result is
   consistent with step 3, so this doesn't appear to be a confound.

5. Compared full get_config(simple_types=True) between HSTUModel(**common_kwargs)
   and SASRecModel(**common_kwargs) — the only remaining structural differences
   are transformer_layers_type (STULayers vs SASRecTransformerLayers) and the
   two HSTU-only flags already ruled out in step 3.

6. Verified data preparation is identical between the two models. Both use
   data_preparator_type=SASRecDataPreparator by default. Ran
   process_dataset_train manually for both models on the same Dataset object
   and compared the resulting interactions.df:
   - Row/column counts matched (3712 rows).
   - Only extra column on the HSTU side is unix_ts (used for relative time
     attention).
   - After restricting to common columns (user_id, item_id, weight, datetime)
     and sorting, hstu_df.equals(sasrec_df) returned True — prepared training
     data is identical.
   - Spot-checked unix_ts for a single user across their full history — it
     correctly varies with datetime (e.g. 2026-04-28 -> 1777334400,
     2026-04-30 -> 1777507200, 2026-08-18 -> 1787011200), so relative-time
     computation is not producing constant/degenerate values.

This chain of eliminations points to something inside STULayers itself
(forward pass or weight initialization) rather than data preparation,
relative-attention flags, or training hyperparameters.

Note: our dataset is proprietary and cannot be shared due to NDA. All numbers
above come from internal tests; the reproduction snippet above uses a
placeholder tiny_train_df. Happy to describe its schema/statistics further or
run additional diagnostic snippets on our data if that would help narrow down
the issue, but cannot share the raw data or notebook itself.

### Expected behavior

train_loss should decrease similarly to SASRecModel under a comparable
configuration, especially in the massively overfit-prone 100-user tiny case
where any reasonably-implemented model should memorize the training data
within 50 epochs.

### Relevant logs and/or screenshots

See loss curves above (train_loss values per epoch, logged via
pytorch_lightning CSVLogger). Unable to share the underlying dataset or
notebook due to NDA restrictions, but can run further targeted diagnostics
on our data and report back specific numbers/behavior on request.

### Operating System

Linux

### Python Version

3.12

### RecTools version

0.19.0
